internal/bcdb/worldstate_query_processor_test.go

```diff
@@ -324,8 +321,296 @@ func TestGetData(t *testing.T) {
+						Metadata: m,
+					},
+					{
+						Key:      "key4",
+						Value:    []byte(`{"attr1":"f","attr2":true,"attr3":"m","attr4":-100}`),
+						Metadata: m,
+					},
+					{
+						Key:      "key5",
+						Value:    []byte(`{"attr1":"g","attr2":true,"attr3":"n","attr4":-101}`),
+						Metadata: m,
+					},
+					{
+						Key:      "key6",
+						Value:    []byte(`{"attr1":"h","attr2":true,"attr3":"o","attr4":-102}`),
+						Metadata: m,
+					},
+				},
+			},
+		}
+
+		indexUpdates, err := stateindex.ConstructIndexEntries(dbsUpdates, db)
+		require.NoError(t, err)
+		for indexDB, updates := range indexUpdates {
+			dbsUpdates[indexDB] = updates
+		}
+		require.NoError(t, db.Commit(dbsUpdates, 3))
+	}
+
+	tests := []struct {
+		name        string
+		dbName      string
+		userID      string
+		query       []byte
+		expectedKVs map[string]*types.KVWithMetadata
+		expectedErr string
+	}{
+		{
+			name:   "fetch records based on boolean matching",
+			dbName: "db1",
+			userID: "user1",
+			query: []byte(
+				`{
+					"selector": {
+						"attr2": {
+							"$eq": true
+						}
+					}
+				}`,
+			),
+			expectedKVs: map[string]*types.KVWithMetadata{
+				"key4": {
+					Key:      "key4",
+					Value:    []byte(`{"attr1":"f","attr2":true,"attr3":"m","attr4":-100}`),
+					Metadata: m,
+				},
+				"key5": {
+					Key:      "key5",
+					Value:    []byte(`{"attr1":"g","attr2":true,"attr3":"n","attr4":-101}`),
+					Metadata: m,
+				},
+				"key6": {
+					Key:      "key6",
+					Value:    []byte(`{"attr1":"h","attr2":true,"attr3":"o","attr4":-102}`),
+					Metadata: m,
+				},
+			},
+		},
+		{
+			name:   "fetch records based on string",
+			dbName: "db1",
+			userID: "user1",
+			query: []byte(
+				`{
+					"selector": {
+						"attr1": {
+							"$gt": "",
+							"$lte": "d"
+						}
+					}
+				}`,
+			),
+			expectedKVs: map[string]*types.KVWithMetadata{
+				"key1": {
+					Key:      "key1",
+					Value:    []byte(`{"attr1":"a","attr2":false,"attr3":"z","attr4":100}`),
+					Metadata: m,
+				},
+				"key2": {
+					Key:      "key2",
+					Value:    []byte(`{"attr1":"b","attr2":false,"attr3":"y","attr4":101}`),
+					Metadata: m,
+				},
+				"key3": {
+					Key:      "key3",
+					Value:    []byte(`{"attr1":"c","attr2":false,"attr3":"x","attr4":102}`),
+					Metadata: m,
+				},
+			},
+		},
+		{
+			name:   "empty result due to acl",
+			dbName: "db1",
+			userID: "user2",
+			query: []byte(
+				`{
+					"selector": {
+						"attr2": {
+							"$eq": true
+						}
+					}
+				}`,
+			),
+		},
+		{
+			name:   "user cannot read from system database",
+			dbName: worldstate.ConfigDBName,
+			userID: "user1",
+			query: []byte(
+				`{
+					"selector": {
+						"attr1": {
+							"$gt": "",
+							"$lte": "d"
+						}
+					}
+				}`,
+			),
+			expectedErr: "no user can directly read from a system database [" + worldstate.ConfigDBName + "]",
+		},
+		{
+			name:   "user does not have read permission",
```

Comment:
another test case: what happens if the query contains an attribute that is not indexed?

Reply:
Why just stop there? why not test the other package again here? When the attribute is not indexed, the queryexecutor would return an error.